File: grind75/Dec21/3.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if not s:
            return 0
        largest = 1
        d = {}
        start, end = 0, 0
        for i, char in enumerate(s):
            if char in d and d[char] >= start:
                start = d[char] + 1
            end = i
            largest = max(largest, end - start + 1)
            d[char] = i
        
        return largest

File: grind75/Dec21/3.go
func lengthOfLongestSubstring(s string) int {
    l := len(s)
    if l == 0 { return l }
    m := map[byte]int{}
    i,j := 0,0
    ret := math.MinInt
    for j < l {
        if m[s[j]] == 0 {
            m[s[j]] += 1
            j++
            continue
        }
        ret = max(ret, j-i)
        for i < j {
            m[s[i]] -= 1
            i++
            if s[i-1] == s[j] {
                break
            }
        }
    }
    ret = max(ret, j-i)
    return ret
}
